### PR TITLE
Clinical trials

### PR DESCRIPTION
**Pull request by Leowisd**
This implements https://github.com/oncokb/oncokb/issues/2479

### CODE REVIEW

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 1]
```diff
@@ -1,242 +1,215 @@
...
+            @ApiResponse(code = 200, message = "OK", responseContainer = "Map"),
+            @ApiResponse(
+                code = 400,
+                message = "Error",
+                response = String.class
+            ),
+        }
+    )
+    @RequestMapping(
+        value = "/trials/cancerType",
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:113]: The api path does not make sense to me. I would rather use `trials-by-cancer-type`

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 2]
```diff
@@ -1,242 +1,215 @@
...
+    )
+    public ResponseEntity<List<ClinicalTrial>> trialsGetByCancerType(
+        @ApiParam(
+            value = "The cancer type that clinical trials belong to. Support cancer type name and OncoTree Code. Example: Glioma or AASTR. Support special cancer types: ALL_TUMORS, ALL_SOLID_TUMORS, ALL_LIQUID_TUMORS."
+        ) @RequestParam(
+            value = "cancerType",
+            required = true
+        ) String cancerType,
+        @ApiParam(
+            value = "Consisted of single/multuple drugs. Support drugs name or drugs' NCIT code. For multiple drugs treatment, use '+' as separator. Example: Everolimus+C95701"
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:125]:
```suggestion
            value = "Consisted of single/multiple drugs. Support drug name or NCIT code. For multiple drugs treatment, use '+' as separator. Example: Binimetinib+Ribociclib"
```

File: web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java [hunk 3]
```diff
@@ -1,242 +1,215 @@
...
+        ) @RequestParam(value = "treatment", required = false) String treatment,
+        @ApiParam(
+            value = "The address of your location. Support zip code. Must be spcified with country. Example: New York City, NY"
+        ) @RequestParam(value = "address", required = false) String address,
+        @ApiParam(
+            value = "The country of your location. Must be specified with address. Example: United States"
+        ) @RequestParam(value = "country", required = false) String country,
+        @ApiParam(
+            value = "The radius from your location. Must be specified with address and country. Example: 100, which means all trails have any site locates within 100 km from your location. If not specify, the default distance is 100km."
```
Comment (zhx828) [inline web/src/main/java/org/mskcc/cbio/oncokb/api/pub/v1/TrialsApiController.java:134]: These descriptions should be put into constant and share with annotation endpoints.

**Review by zhx828 — CHANGES_REQUESTED** [review]
the .json.gz should not be included in the repo.

File: core/src/main/java/org/mskcc/cbio/oncokb/util/ClinicalTrialsUtils.java [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+            }
+        }
+
+        return trials;
+    }
+
+    public List<ClinicalTrial> getTrialsByIDList(List<String> ids) {
+        return ids
+            .stream()
+            .map(id -> getAllTrials().get(id))
```
[the remark this answers is not in the document]
Comment (Leowisd) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/ClinicalTrialsUtils.java:341]: I think each trial should have nctId from the pipeline

File: core/src/main/java/org/mskcc/cbio/oncokb/util/CacheUtils.java [hunk 5]
```diff
@@ -120,6 +129,39 @@ public void update(Observable o, Object arg) {
...
+    private static Observer clinicalTrialsObserver = new Observer(){
+        @Override
+        public void update(Observable o, Object arg){
+            try {
+                clinicalTrials = ClinicalTrialsUtils.getInstance().loadTrialsMap();
+            } catch (IOException | ParseException e) {
+                e.printStackTrace();
+            }
+        }
+    };
```
[the remark this answers is not in the document]
Comment (Leowisd) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/CacheUtils.java:163]: It makes sense. Remove the observer of the clinical trials.

File: core/src/main/java/org/mskcc/cbio/oncokb/util/ClinicalTrialsUtils.java [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+                os
+            );
+        }
+        result =
+            new Gson()
+            .fromJson(
+                    os.toString("UTF-8"),
+                    new TypeToken<Map<String, Tumor>>() {}.getType()
+                );
+        os.close();
```
[the remark this answers is not in the document]
Comment (Leowisd) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/ClinicalTrialsUtils.java:77]: https://stackoverflow.com/questions/14503881/strange-behavior-when-deserializing-nested-generic-classes-with-gson/14506181#14506181

File: core/src/main/java/org/mskcc/cbio/oncokb/util/GzipUtils.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.zip.GZIPInputStream;
+import java.util.zip.GZIPOutputStream;
+
+/**
+ * Created by Yifu Yao on 3/12/2021
+ */
+public class GzipUtils {
```
[the remark this answers is not in the document]
Comment (Leowisd) [reply core/src/main/java/org/mskcc/cbio/oncokb/util/GzipUtils.java:17]: Don't find specific library util to use.

**Review by zhx828 — CHANGES_REQUESTED** [review]
We need to prefill the coordinates for all sites in the sites.json For anyone without/or cannot find coordinates, we need to give some defaults(maybe -1,-1, or do not search the location if the coordinates null) to let the script knows it has been searched. Otherwise, every time you make a search, the script looks for locations for the ones without the coordinates.

**Comment by zhx828** [comment]
@Leowisd could you give me a list of distinct recruitmentStatus? I'm wondering whether we should filter out the sites without recruiting.

**Comment by zhx828** [comment]
@Leowisd could you also check why sites of NCT02799095 do not include any locations outside of united states? It's a match  when I search toronto/canada/MEL

**Comment by zhx828** [comment]
@Leowisd also in the sites.json, the recruitment_status/recruitment_status_date will be different trial from trial. The sites.json should not include this info.

**Comment by Leowisd** [comment]
> @Leowisd could you give me a list of distinct recruitmentStatus? I'm wondering whether we should filter out the sites without recruiting.

**Comment by Leowisd** [comment]
> @Leowisd could you also check why sites of NCT02799095 do not include any locations outside of united states? It's a match when I search toronto/canada/MEL

ACTIVE
INACTIVE 
APPROVED  
COMPLETED 
ADMINISTRATIVELY_COMPLETE 
CLOSED_TO_ACCRUAL 
TEMPORARILY_CLOSED_TO_ACCRUAL 
CLOSED_TO_ACCRUAL_AND_INTERVENTION 
WITHDRAWN

File: core/src/test/java/org/mskcc/cbio/oncokb/util/ClinicalTrialsUtilsTest.java [hunk 8]
```diff
@@ -126,15 +126,25 @@ public void testFilterTrialsByDrugNameOrCode() {
...
-                "Columbia, MO, United States",
+                "toronto, Canada",
                 100.0
             );
-        assertTrue("Filter by location failed", res.size() != 0);
+        assertTrue("Filter by location failed", res.size() != 0);       
+        Gson gson = new GsonBuilder().setDateFormat("yyyy-MM-dd").create();
+        try {
+            FileWriter file = new FileWriter(
+                "C:\\Users\\Yifu\\Desktop\\test.json"
```
Comment (zhx828) [inline core/src/test/java/org/mskcc/cbio/oncokb/util/ClinicalTrialsUtilsTest.java:141]: Does it have to be hard coded?

**Review by zhx828 — CHANGES_REQUESTED** [review]

**Review by zhx828 — APPROVED** [review]

**Pull request merged**